File: tests/test_tui_auto_scroll.c
/*
 * test_tui_auto_scroll.c - Unit tests for TUI auto-scroll logic
 * Tests the auto-scroll condition: scroll_offset >= max_scroll - 1 (98-100% range)
 *
 * This test focuses on the auto-scroll logic without full TUI dependencies
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

// Test framework colors
#define TEST_COLOR_RESET "\033[0m"
#define TEST_COLOR_GREEN "\033[32m"
#define TEST_COLOR_RED "\033[31m"
#define TEST_COLOR_YELLOW "\033[33m"
#define TEST_COLOR_CYAN "\033[36m"

// Test counters
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

// Test utilities
static void print_test_result(const char *test_name, int passed) {
    tests_run++;
    if (passed) {
        tests_passed++;
        printf(TEST_COLOR_GREEN "✓ PASS" TEST_COLOR_RESET " %s\n", test_name);
    } else {
        tests_failed++;
        printf(TEST_COLOR_RED "✗ FAIL" TEST_COLOR_RESET " %s\n", test_name);
    }
}

static void print_summary(void) {
    printf("\n" TEST_COLOR_CYAN "=== Test Summary ===" TEST_COLOR_RESET "\n");
    printf("Tests run:    %d\n", tests_run);
    printf("Tests passed: %d\n", tests_passed);
    printf("Tests failed: %d\n", tests_failed);

    if (tests_failed == 0) {
        printf(TEST_COLOR_GREEN "\n✓ All tests passed!" TEST_COLOR_RESET "\n");
    } else {
        printf(TEST_COLOR_RED "\n✗ Some tests failed!" TEST_COLOR_RESET "\n");
    }
}

// Mock TUI modes (from tui.h)
typedef enum {
    TUI_MODE_INSERT,
    TUI_MODE_NORMAL,
    TUI_MODE_COMMAND
} TUIMode;

// Function to test auto-scroll logic
static int should_auto_scroll_normal_command(int scroll_offset, int max_scroll, int content_lines) {
    // This replicates the logic from tui.c lines ~1750-1770
    // Auto-scroll logic for NORMAL/COMMAND mode

    if (content_lines == 0 || max_scroll <= 0) {
        // No content or everything fits in viewport
        return 1;
    } else if (scroll_offset >= max_scroll - 1) {
        // Already at bottom (with 1-line tolerance for 98-100% range)
        return 1;
    }
    return 0;
}

// Test cases
static void test_auto_scroll_at_bottom(void) {
    // Test case 1: At bottom (scroll_offset == max_scroll)
    int result = should_auto_scroll_normal_command(100, 100, 200);
    print_test_result("At bottom (100/100) should auto-scroll", result == 1);
}

static void test_auto_scroll_one_line_from_bottom(void) {
    // Test case 2: One line from bottom (scroll_offset == max_scroll - 1)
    int result = should_auto_scroll_normal_command(99, 100, 200);
    print_test_result("One line from bottom (99/100) should auto-scroll", result == 1);
}

static void test_auto_scroll_two_lines_from_bottom(void) {
    // Test case 3: Two lines from bottom (scroll_offset == max_scroll - 2)
    int result = should_auto_scroll_normal_command(98, 100, 200);
    print_test_result("Two lines from bottom (98/100) should NOT auto-scroll", result == 0);
}

static void test_auto_scroll_far_from_bottom(void) {
    // Test case 4: Far from bottom
    int result = should_auto_scroll_normal_command(50, 100, 200);
    print_test_result("Far from bottom (50/100) should NOT auto-scroll", result == 0);
}

static void test_auto_scroll_small_max_scroll(void) {
    // Test case 5: Edge case with small max_scroll
    int result = should_auto_scroll_normal_command(1, 1, 10);
    print_test_result("Small max_scroll (1/1) should auto-scroll", result == 1);
}

static void test_auto_scroll_zero_max_scroll(void) {
    // Test case 6: Edge case with zero max_scroll (all content fits)
    int result = should_auto_scroll_normal_command(0, 0, 5);
    print_test_result("Zero max_scroll (0/0, content fits) should auto-scroll", result == 1);
}

static void test_auto_scroll_negative_scroll(void) {
    // Test case 7: Negative scroll offset (shouldn't happen in practice)
    int result = should_auto_scroll_normal_command(-1, 10, 20);
    print_test_result("Negative scroll offset should NOT auto-scroll", result == 0);
}

static void test_auto_scroll_no_content(void) {
    // Test case 8: No content (content_lines == 0)
    int result = should_auto_scroll_normal_command(0, 0, 0);
    print_test_result("No content should auto-scroll", result == 1);
}

static void test_auto_scroll_fits_in_viewport(void) {
    // Test case 9: Everything fits in viewport (max_scroll <= 0)
    // When content_lines <= viewport_height, max_scroll would be 0 or negative
    int result = should_auto_scroll_normal_command(0, -5, 10);
    print_test_result("Content fits in viewport should auto-scroll", result == 1);
}

static void test_auto_scroll_insert_mode_always(void) {
    // Test case 10: INSERT mode always auto-scrolls
    // This is tested by checking that the condition doesn't matter for INSERT mode
    // In the actual code, INSERT mode calls window_manager_scroll_to_bottom directly
    print_test_result("INSERT mode always auto-scrolls (by design)", 1);
}

// Test the actual percentage calculation (from tui.c line ~181)
static void test_percentage_calculation(void) {
    // Test the percentage calculation: (scroll_offset * 100 + max_scroll / 2) / max_scroll
    int scroll_offset = 75;
    int max_scroll = 100;
    int percentage = (scroll_offset * 100 + max_scroll / 2) / max_scroll;

    print_test_result("Percentage calculation (75/100 = 75%)", percentage == 75);

    // Test rounding
    scroll_offset = 99;
    max_scroll = 100;
    percentage = (scroll_offset * 100 + max_scroll / 2) / max_scroll;
    print_test_result("Percentage rounding (99/100 = 99%)", percentage == 99);

    // Test edge case with max_scroll = 0 (should not divide by zero in actual code)
    scroll_offset = 0;
    max_scroll = 0;
    // In actual code, this is protected by: if (max_scroll > 0)
    print_test_result("Zero max_scroll handled safely in percentage calc", 1);
}

// Test boundary conditions
static void test_boundary_conditions(void) {
    printf(TEST_COLOR_YELLOW "\n=== Boundary Condition Tests ===" TEST_COLOR_RESET "\n");

    // Test exactly at the boundary: scroll_offset = max_scroll - 1
    int result = should_auto_scroll_normal_command(99, 100, 200);
    print_test_result("Boundary: scroll_offset = max_scroll - 1 should auto-scroll", result == 1);

    // Test just below boundary: scroll_offset = max_scroll - 2
    result = should_auto_scroll_normal_command(98, 100, 200);
    print_test_result("Boundary: scroll_offset = max_scroll - 2 should NOT auto-scroll", result == 0);

    // Test large numbers
    result = should_auto_scroll_normal_command(9999, 10000, 20000);
    print_test_result("Large numbers: 9999/10000 should auto-scroll", result == 1);

    result = should_auto_scroll_normal_command(9998, 10000, 20000);
    print_test_result("Large numbers: 9998/10000 should NOT auto-scroll", result == 0);
}

int main(void) {
    printf(TEST_COLOR_CYAN "=== TUI Auto-Scroll Logic Tests ===" TEST_COLOR_RESET "\n");
    printf("Testing auto-scroll condition: scroll_offset >= max_scroll - 1 (98-100%% range)\n\n");

    // Run tests
    test_auto_scroll_at_bottom();
    test_auto_scroll_one_line_from_bottom();
    test_auto_scroll_two_lines_from_bottom();
    test_auto_scroll_far_from_bottom();
    test_auto_scroll_small_max_scroll();
    test_auto_scroll_zero_max_scroll();
    test_auto_scroll_negative_scroll();
    test_auto_scroll_no_content();
    test_auto_scroll_fits_in_viewport();
    test_auto_scroll_insert_mode_always();
    test_percentage_calculation();
    test_boundary_conditions();

    print_summary();

    return tests_failed == 0 ? 0 : 1;
}

File: zig/tests/test_tui_auto_scroll.zig
//! tests/test_tui_auto_scroll.zig — Zig port of tests/test_tui_auto_scroll.c
//!
//! Tests the TUI auto-scroll condition:
//!   should_auto_scroll  ⟺  scroll_offset >= max_scroll - 1
//!
//! All logic is pure arithmetic — no ncurses required.

const std = @import("std");

// ---------------------------------------------------------------------------
// Replicated auto-scroll predicate (mirrors tui.c NORMAL/COMMAND path)
// ---------------------------------------------------------------------------

/// Returns true when the conversation view should auto-scroll to the bottom.
///
/// Mirrors the logic from tui.c (NORMAL/COMMAND mode branch):
///   - If there is no content, or everything fits in the viewport → scroll.
///   - If we are within one line of the bottom → scroll.
///   - Otherwise → don't scroll (user has scrolled up deliberately).
fn shouldAutoScroll(scroll_offset: i32, max_scroll: i32, content_lines: i32) bool {
    if (content_lines == 0 or max_scroll <= 0) {
        return true;
    }
    return scroll_offset >= max_scroll - 1;
}

// ---------------------------------------------------------------------------
// Basic cases
// ---------------------------------------------------------------------------

test "auto-scroll: at bottom (100/100) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(100, 100, 200));
}

test "auto-scroll: one line from bottom (99/100) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(99, 100, 200));
}

test "auto-scroll: two lines from bottom (98/100) should NOT auto-scroll" {
    try std.testing.expect(!shouldAutoScroll(98, 100, 200));
}

test "auto-scroll: far from bottom (50/100) should NOT auto-scroll" {
    try std.testing.expect(!shouldAutoScroll(50, 100, 200));
}

// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------

test "auto-scroll: max_scroll == 1, at bottom (1/1) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(1, 1, 10));
}

test "auto-scroll: max_scroll == 0 (all content fits) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(0, 0, 5));
}

test "auto-scroll: negative scroll offset should NOT auto-scroll" {
    try std.testing.expect(!shouldAutoScroll(-1, 10, 20));
}

test "auto-scroll: no content (content_lines == 0) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(0, 0, 0));
}

test "auto-scroll: negative max_scroll (content fits in viewport) should auto-scroll" {
    try std.testing.expect(shouldAutoScroll(0, -5, 10));
}

test "auto-scroll: INSERT mode always auto-scrolls by design (placeholder)" {
    // INSERT mode calls window_manager_scroll_to_bottom directly, bypassing
    // this predicate entirely.  The logic is correct by construction.
    try std.testing.expect(true);
}

// ---------------------------------------------------------------------------
// Boundary conditions
// ---------------------------------------------------------------------------

test "auto-scroll boundary: scroll_offset == max_scroll - 1 → auto-scroll" {
    try std.testing.expect(shouldAutoScroll(99, 100, 200));
}

test "auto-scroll boundary: scroll_offset == max_scroll - 2 → no auto-scroll" {
    try std.testing.expect(!shouldAutoScroll(98, 100, 200));
}

test "auto-scroll boundary: large numbers 9999/10000 → auto-scroll" {
    try std.testing.expect(shouldAutoScroll(9999, 10000, 20000));
}

test "auto-scroll boundary: large numbers 9998/10000 → no auto-scroll" {
    try std.testing.expect(!shouldAutoScroll(9998, 10000, 20000));
}

// ---------------------------------------------------------------------------
// Percentage calculation (mirrors tui.c status-bar display logic)
// ---------------------------------------------------------------------------

/// Integer percentage with rounding-by-adding-half-divisor.
fn scrollPercent(scroll_offset: i32, max_scroll: i32) i32 {
    if (max_scroll <= 0) return 100;
    return @divTrunc(scroll_offset * 100 + @divTrunc(max_scroll, 2), max_scroll);
}

test "scroll percent: 75 / 100 == 75%" {
    try std.testing.expectEqual(@as(i32, 75), scrollPercent(75, 100));
}

test "scroll percent: 99 / 100 == 99%" {
    try std.testing.expectEqual(@as(i32, 99), scrollPercent(99, 100));
}

test "scroll percent: 0 / 0 returns 100% (guard)" {
    try std.testing.expectEqual(@as(i32, 100), scrollPercent(0, 0));
}

test "scroll percent: 0 / 100 == 0%" {
    try std.testing.expectEqual(@as(i32, 0), scrollPercent(0, 100));
}

test "scroll percent: 50 / 100 == 50%" {
    try std.testing.expectEqual(@as(i32, 50), scrollPercent(50, 100));
}
